pheps/phep-9999.md

```diff
@@ -0,0 +1,137 @@
+```
+PHEP: 9999
+Title: PyHC standardization for Python time objects
+Author: Nabil Freij <nabil.freij@gmail.com> <https://orcid.org/0000-0002-6253-082X>
+Discussions-To: https://github.com/heliophysicsPy/standards/pull/X
+Revision: 1
+Status: Draft
+Type: Standards Track
+Content-Type: text/markdown; charset=UTF-8; variant=CommonMark
+Created: 08-July-2024
+Post-History: 08-July-2024
+```
+
+# Abstract
+
+This PHEP recommends that all projects across the PyHC ecosystem use the standard library [datetime](https://docs.python.org/3/library/datetime.html) module or if the project has more complex requirements, they should use `astropy.time.Time` instead of creating their own time objects.
+
+Eventually, we want to encourage that all PyHC libraries allow `astropy.time.Time` as valid time inputs to their libraries.
+This has some roadblocks currently which will not allow this to happen in the near future, but hopefully with the PyHC community behind this PHEP, we can push for better astropy integration with the broader Scientific Python ecosystem.
+
+Any existing projects that have their own time object are strongly encouraged to replace their custom time objects with `astropy.time.Time`.
+
+# Motivation
+
+One of the strategic goals of the PyHC Project is to "Coordinate development across projects to minimize duplication of effort".
+
+However, all of the core (and non-core) PyHC packages were created not only before the existence of PyHC, but also before many core libraries existed within the wider Scientific Python ecosystem (from `numpy` to `scipy` to `astropy` to `spiceypy` etc).
+This meant that a some PyHC packages have had to create their own implementations of specific components like time or coordinates due to necessity.
+
+Now that the upstream ecosystem has grown at a rapid pace, the need for PyHC package specific implementations has reduced considerably.
+There are at least three areas ("time", "coordinates" and "units") that require standardization, however the focus for this PHEP is to start at "time", as that is considered the *simplest* case for standardization.
+
+To try and address the need to reduce duplication and improve interoperability between PyHC packages, decisions have to be made about reducing the number of ways the community handles the same type of operation but it has to be considered within the wider Scientific Python ecosystem.
+
+For example, the reason that Astropy's time framework was created was to add support for astronomical formats (e.g., Julian Date (JD), Modified JD (MJD) and precise timing (e.g., a nanosecond over a Hubble Time).
+None of which is possible when using the [datetime](https://docs.python.org/3/library/datetime.html) module.
+
```

Comment:
```suggestion

# Use case

Several independent researchers are attempting to match coronal features during an eruption observed by the Atmospheric Imaging Assembly (AIA) on the Solar Dynamics Observatory (SDO) to in situ measurements observed by Parker Solar Probe (PSP). Among these researchers are summer student interns who are using PyHC packages for the first time. This study requires careful comparisons of the times of the remote and in situ observations. In particular, the researchers must make numerous time comparisons, while taking into account transit time and spatial differences. 

The researchers decide to use SunPy (which makes use of `astropy.time.Time`) to read in and analyze remote observations from SDO/AIA. They then decide to either use SpacePy (and its `TickTock` class) or PySPEDAS (which uses `datetime64`?) to read in and analyze in situ observations from PSP. However, operations like `Time(...) - TickTock(...)` result in an exception(?).

Currently, no functionality exists(?) among PyHC packages to convert between or directly compare `TickTock`, `Time`, and `datetime64` objects. This lack of interoperability has the following consequences:

 - Students and researchers would need to learn how to use and interact with multiple different time objects, which increases the onboarding time for student interns.
 - Each researcher would need to write or acquire functionality that would convert the different time classes into a common class.(?)
 - The effort required to perform this analysis would be increased.
 - User experience is degraded.

If the ecosystem were to standardize on a common time framework (or at least one with a common API), the consequences would be:

 - Cross-disciplinary researchers would only need to learn a single API for dealing with time.
 - Time delta operations would be greatly simplified: subtracting a time object from another would produce a consistent time delta object.
 - The effort needed to perform this analysis would decrease.
 - The likelihood of unnecessary software duplication is reduced.
 - User experience when working with multiple PyHC packages would be improved.
 - The onboarding time for students would be reduced, and the suitability of this topic as a research project for student interns would be increased.
 - `astropy.time.TimeDelta` objects can be converted to `astropy.units.Quantity` objects using the `.to` method, which would enable multiplication with a velocity to get a distance.

```
Would it make sense to spell out some [use cases](https://en.wikipedia.org/wiki/Use_case) to show more about why we're hoping to standardize around time?  I attempted to draft one here, with question marks about the parts I'm uncertain about.

Reply:
Hi @namurphy - I suggest trying the experiment. 

It seems that we have a proposed standard and are looking for hypothetical problems to justify it. All the packages understand `datetime`, and the hypothetical does not indicate exactly where `datetime` is insufficient. I expect one could come up with a concrete example that the standard would have fixed, but it will likely be an uncommon use case - if so, we should debate if the burden of the standard on everyone is worth simplifying an uncommon use case. 

I'd really like to see the participants try to use other packages together and study the package's source code. Five ideas for standards or cross-cutting projects will probably result, and I doubt this one involving `AstroPy.Time` will be one of them. I am very reluctant to propose a standard based on hypothetical or contrived problems with claims that may prove wrong if one tries to solve the problem. There should be a PHEP for time, but there should not be speculation to justify it. Also, the implications should be researched (again, without speculations).

**Much more research is needed. I don't see the rush to have this voted on.**

I've started with an ISO 8601 timestamp of interest and used SpacePy, SunPy, PySPEDAS, and SpiceyPy to do a coordinate transform. There may have been a time standard that would have simplified things for me - if I could have bypassed the custom time objects and just passed an ISO 8601 string to the transform functions, things would have been much easier. In my problem, there was no need to expose the package's time abstraction to the user. Several packages require the user to understand custom time and/or coordinate objects (or don't allow dimensionless vectors) to do something conceptually simple - pass a timestamp and a vector and get a new vector. I don't know if there is an actual proposal here - what is there to say beyond "make simple things easy to do" or "don't require the scientist to understand a software abstraction if unnecessary?"

As a result of this experience, I submitted an issue requesting a keyword in the `hapiclient` function to request the output as `datetimes` instead of the default of the raw strings served by the HAPI server. There is no reason for the general user to learn and use a separate function, `hapitime2datetime`, to convert from the raw ISO 8601 strings to `datetimes`. In the next major release, we'll make this the default behavior because this is what most want.